### PR TITLE
HF Buckets integration to Trainer

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?

This PR adds support for HF Buckets in Trainer. Instead of storing checkpoints locally or on the hub repository, we can leverage HF Buckets for that (S3 like storage + XET dedup). 

With buckets (`push_to_buckets=True`), users don't need to save the checkpoints on the hub anymore like before (hub_strategy="checkpoint" or "all_checkpoints") as we push all checkpoints to the bucket. Still, pushing to the hub is still useful (hub_strategy="every_save") as it will upload a version of a model that can we load with transformers. 

Features:
- Saving checkpoints to HF Buckets
- Resuming training from HF Buckets

Right now, everything should be pretty much compatible like before since we are just synching a local dir to the HF Buckets async but in the future, it would be a lot better if  we are able load / save directly from / to HF Buckets without going through the disk, might be useful for deepspeed and fsdp.  

## Usage

```python 
from transformers import Trainer, TrainingArguments

trainer = Trainer(
    model=model,
    args=TrainingArguments(
        output_dir="out",
        save_strategy="steps",
        push_to_bucket=True,
        bucket_id="bucket_id",
    ),
    train_dataset=train_dataset,
)
trainer.train()

# resume on a fresh machine — pulls the latest checkpoint from the bucket
trainer.train(resume_from_checkpoint="hf://buckets/my-org/my-run")
```

### CODE REVIEW

File: docs/source/en/trainer_recipes.md [hunk 1]
```diff
@@ -247,3 +247,33 @@ trainer.train(resume_from_checkpoint="out/checkpoint-1000")
 When resuming, [`Trainer`] restores the optimizer state, scheduler state, and RNG state.
 
 Checkpoint resuming requires optimizer and scheduler state files in the checkpoint directory. If those files are missing (for example, when `save_only_model=True`), the optimizer restarts from scratch.
+
+### Checkpointing to a bucket
+
+Set `push_to_bucket=True` to sync checkpoints to a [Hugging Face bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets) instead of (or alongside) keeping them only on local disk. A bucket is mutable, Xet-backed object storage with no git history — a better fit for the rotating, throwaway nature of intermediate checkpoints than a model repository. Only changed files (and, within them, changed chunks) are transferred on each save.
```
Comment (stevhliu) [inline docs/source/en/trainer_recipes.md]:
```suggestion
Set `push_to_bucket=True` to sync checkpoints to a [Hugging Face bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets). Checkpoints still save to local disk, but the bucket receives a copy of each one. A bucket is mutable, Xet-backed object storage with no git history. Buckets suit intermediate checkpoints better than a model repository, since checkpoints rotate constantly and are meant to be thrown away. Each save transfers only the changed files, and only the changed chunks within them.
```

File: docs/source/en/trainer_recipes.md [hunk 2]
```diff
@@ -247,3 +247,33 @@ trainer.train(resume_from_checkpoint="out/checkpoint-1000")
...
+        push_to_bucket=True,
+        bucket_id="my-org/my-run",
+        ...
+    ),
+    train_dataset=train_dataset,
+)
+trainer.train()
+```
+
+Each checkpoint is uploaded to the bucket under its own `checkpoint-<step>/` prefix and **kept there even after it is rotated away locally** by [`~TrainingArguments.save_total_limit`] (i.e. the bucket accumulates every checkpoint). Uploads run asynchronously, so they don't block training.
```
Comment (stevhliu) [inline docs/source/en/trainer_recipes.md]:
```suggestion
[`Trainer`] uploads each checkpoint to the bucket under its own `checkpoint-<step>/` prefix and keeps it there even after [`~TrainingArguments.save_total_limit`] rotates it off local disk.. The bucket accumulates every checkpoint. Uploads run asynchronously, so they don't block training.
```

File: docs/source/en/trainer_recipes.md [hunk 3]
```diff
@@ -247,3 +247,33 @@ trainer.train(resume_from_checkpoint="out/checkpoint-1000")
...
+
+`push_to_bucket` is independent of `push_to_hub`: use it on its own to store checkpoints without creating a model repository, or together with `push_to_hub` to also publish the model.
```
Comment (stevhliu) [inline docs/source/en/trainer_recipes.md]:
```suggestion
`push_to_bucket` is independent of `push_to_hub`. Use it on its own to store checkpoints without creating a model repository, or together with `push_to_hub` to also publish the model. 
```

File: docs/source/en/trainer_recipes.md [hunk 4]
```diff
@@ -247,3 +247,33 @@ trainer.train(resume_from_checkpoint="out/checkpoint-1000")
...
+    train_dataset=train_dataset,
+)
+trainer.train()
+```
+
+Each checkpoint is uploaded to the bucket under its own `checkpoint-<step>/` prefix and **kept there even after it is rotated away locally** by [`~TrainingArguments.save_total_limit`] (i.e. the bucket accumulates every checkpoint). Uploads run asynchronously, so they don't block training.
+
+`push_to_bucket` is independent of `push_to_hub`: use it on its own to store checkpoints without creating a model repository, or together with `push_to_hub` to also publish the model. 
+
+To resume from a bucket, pass its handle to `resume_from_checkpoint`. [`Trainer`] downloads the checkpoint into `output_dir` and continues from it — handy for resuming on a fresh machine where nothing is on local disk yet. Pass the bucket root to resume from its latest checkpoint, or a full checkpoint handle to resume from a specific one:
```
Comment (stevhliu) [inline docs/source/en/trainer_recipes.md]:
```suggestion
To resume from a bucket, pass its handle to `resume_from_checkpoint`. [`Trainer`] downloads the checkpoint into `output_dir` and continues from it. This helps when resuming on a fresh machine where nothing is on local disk yet. Pass the bucket root to resume from its latest checkpoint, or a full checkpoint handle to resume from a specific one:
```

**Review by stevhliu — APPROVED** [review]
super nice! 🪣

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -1204,6 +1204,22 @@ class TrainingArguments:
             "help": "The revision to use when pushing to the Hub. Can be a branch name, a tag, or a commit hash."
         },
     )
+    push_to_bucket: bool = field(
+        default=False,
+        metadata={
+            "help": "Whether to sync intermediate checkpoints to a Hub bucket (mutable object storage, no git "
+            "history, chunk-level xet dedup) instead of committing them to the model repo. Independent of "
+            "`push_to_hub`; if both are set, the final checkpoint is also committed to the repo at the end of training."
```
Comment (qgallouedec) [inline src/transformers/training_args.py]: nit, but more accurate imo

```suggestion
            "`push_to_hub`; if both are set to `True`, the final checkpoint is also committed to the repo at the end of training."
```

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -4111,60 +4145,89 @@ def _push_from_checkpoint(self, checkpoint_folder: str) -> None:
...
-                repo_id=self.hub_model_id,
-                folder_path=checkpoint_folder,
-                path_in_repo=path_in_repo,
-                commit_message=commit_message + ", checkpoint",
-                token=self.args.hub_token,
-                run_as_future=True,
-                revision=self.args.hub_revision,
+
+            # Full checkpoint -> repo (unchanged behavior, gated by hub_strategy).
+            if self.args.hub_strategy in [HubStrategy.CHECKPOINT, HubStrategy.ALL_CHECKPOINTS]:
```
Comment (qgallouedec) [inline src/transformers/trainer.py:4207]: we don't need the `modeling_files` in this case?

File: docs/source/en/trainer_recipes.md [hunk 7]
```diff
@@ -247,3 +247,33 @@ trainer.train(resume_from_checkpoint="out/checkpoint-1000")
...
+Set `push_to_bucket=True` to sync checkpoints to a [Hugging Face bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets). Checkpoints still save to local disk, but the bucket receives a copy of each one. A bucket is mutable, Xet-backed object storage with no git history. Buckets suit intermediate checkpoints better than a model repository, since checkpoints rotate constantly and are meant to be thrown away. Each save transfers only the changed files, and only the changed chunks within them.
+
+```py
+from transformers import Trainer, TrainingArguments
+
+trainer = Trainer(
+    model=model,
+    args=TrainingArguments(
+        push_to_bucket=True,
+        bucket_id="my-org/my-run",
```
Comment (ArthurZucker) [inline docs/source/en/trainer_recipes.md]: we could have a good default for this IMO, the least amount of args the better!

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -573,8 +575,18 @@ def __init__(
 
         # ---- 9. Hub & output ---------------------------------------------------------
         self.hub_model_id = None  # Set by init_hf_repo() when push_to_hub is enabled
+        self._bucket_id = None  # Canonical id of the checkpoint bucket when push_to_bucket is enabled
+        self.push_in_progress = None  # Tracks the in-flight repo push
         if self.args.push_to_hub:
             self.init_hf_repo()
+        if self.args.push_to_bucket and self.is_world_process_zero():
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:585]: we could have `push_in_progress = True` default to push to a bucket?

**Review by ArthurZucker — APPROVED** [review]
Nice! was wondering if mounting the bucket and just writing to it would not make more sense? 

+ good defaults?! since anyways

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -4102,7 +4136,7 @@ def push_to_hub(
         )
 
     def _push_from_checkpoint(self, checkpoint_folder: str) -> None:
-        """Push model and checkpoint files to the Hub from a checkpoint folder."""
+        """Push model files to the repo and/or sync the checkpoint to the bucket, from a checkpoint folder."""
         if not self.is_world_process_zero() or self.args.hub_strategy == HubStrategy.END:
             return
         # If we haven't finished the last push, we don't do this one unless args.hub_always_push=True.
```
Comment (qgallouedec) [inline src/transformers/trainer.py:4149]: I think this is problematic: when `push_to_bucket=True`, `push_to_hub=False`, `hub_strategy="end"`, this disables the bucket entirely. Would someone use `push_to_hub=False` and `hub_strategy="end"`? Maybe it could be

```diff
- self.args.hub_strategy == HubStrategy.END
+ (self.args.push_to_hub and self.args.hub_strategy == HubStrategy.END) and not push_to_bucket
```

something like this

**Review by qgallouedec — APPROVED** [review]
lgtm, just a few question and edge cases

File: docs/source/en/trainer_recipes.md (hunk 7 above)
Comment (SunMarc) [reply docs/source/en/trainer_recipes.md]: Yeah, I didn't precise but it will default to `hub_model_id` if it is set or default to `output_dir` otherwise.

Comment (SunMarc) [reply docs/source/en/trainer_recipes.md]: so this is actually optional

File: src/transformers/trainer.py (hunk 8 above)
Comment (SunMarc) [reply src/transformers/trainer.py:585]: this is a private attribute not related to push to a bucket, more an arg to track that we are still pushing a checkpoint so we shouldn't interrupt it. If we want to force pushing to a bucket, we would have to set push_to_bucket = True as the default

File: src/transformers/trainer.py (hunk 6 above)
Comment (SunMarc) [reply src/transformers/trainer.py:4207]: There are three things happening sequentially: 
1) if `push to hub`: we update the output_dir to have the most recent model + tokenizer and so on (this is the modeling_files) and we upload that to the hub

2) if `push to hub` +`hub_strategy="checkpoint" or "all_checkpoints", we either update last_checkpoint folder with the new checkpoint or upload the new checkpoint file

3) if `push_to_bucket`, we sync the output_dir folder with the bucket 

One thing that I didn't add but we should maybe is that right now we don't update the `output_dir` with `modeling_files` when pushing to bucket and this is actually something we should maybe to do have parity with push to hub and not create confusion.

File: src/transformers/trainer.py (hunk 9 above)
Comment (SunMarc) [reply src/transformers/trainer.py:4149]: Nice thanks for noticing this. I will update this !

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -1339,9 +1351,12 @@ def train(
 
         Args:
             resume_from_checkpoint (`str` or `bool`, *optional*):
-                If a `str`, local path to a saved checkpoint as saved by a previous instance of [`Trainer`]. If a
-                `bool` and equals `True`, load the last checkpoint in *args.output_dir* as saved by a previous instance
-                of [`Trainer`]. If present, training will resume from the model/optimizer/scheduler states loaded here.
+                If a `str`, local path to a saved checkpoint as saved by a previous instance of [`Trainer`]. A bucket
+                handle resumes from a [bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets) instead:
+                `"hf://buckets/namespace/name"` downloads the latest checkpoint in that bucket, and
```
Comment (lewtun) [inline src/transformers/trainer.py:1360]: Can the string be a nested path in a bucket? 

For example, suppose I am running N experiments on top of the same base model and want all checkpoints to live in the same bucket, I would have something like `hf://buckets/namespace/name/expt-1/checkpoint-500/` and `hf://buckets/namespace/name/expt-2/checkpoint-500` etc

I know that historically we've treated 1 model repo = 1 training run, but for buckets the use case is different and closer to S3 where one does indeed dump checkpoints from multiple experiments in the _same_ bucket.

Comment (SunMarc) [reply src/transformers/trainer.py:1360]: We sync the `local_dir` where the checkpoints are to `hf://buckets/namespace/name`. So if your `local_dir` have the following structure: `expt-1/checkpoint-500/` , `expt-2/checkpoint-500/`, it will work ! 

In the case of Trainer, since we save the checkpoint directly on `output_dir` then we sync on the bucket, it might not work so well. Let me try to see if I can fix this.

Comment (SunMarc) [reply src/transformers/trainer.py:1360]: Pushed a new commit, now you can specify a path in the `bucket_id` like `hf://buckets/namespace/name/exp-1` that will be sync with the content in `output_dir` ! This should help managing all the different experiments.

Comment (lewtun) [reply src/transformers/trainer.py:1360]: thanks this looks great!
